src/main/java/org/littleshoot/proxy/extras/SelfSignedSslEngineSource.java

```diff
@@ -4,43 +4,71 @@
 import java.io.FileInputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.lang.ref.SoftReference;
+import java.math.BigInteger;
 import java.security.KeyStore;
+import java.security.PrivateKey;
+import java.security.PublicKey;
 import java.security.Security;
-import java.security.cert.CertificateException;
-import java.security.cert.X509Certificate;
+import java.security.cert.Certificate;
 import java.util.Arrays;
+import java.util.Enumeration;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
 
 import javax.net.ssl.KeyManager;
 import javax.net.ssl.KeyManagerFactory;
 import javax.net.ssl.SSLContext;
 import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLSession;
 import javax.net.ssl.TrustManager;
 import javax.net.ssl.TrustManagerFactory;
-import javax.net.ssl.X509TrustManager;
+import javax.security.cert.X509Certificate;
 
 import org.apache.commons.io.IOUtils;
 import org.littleshoot.proxy.SslEngineSource;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import sun.security.x509.AlgorithmId;
+import sun.security.x509.CertificateAlgorithmId;
+import sun.security.x509.CertificateExtensions;
+import sun.security.x509.CertificateIssuerName;
+import sun.security.x509.CertificateSerialNumber;
+import sun.security.x509.CertificateSubjectName;
+import sun.security.x509.CertificateVersion;
+import sun.security.x509.CertificateX509Key;
+import sun.security.x509.Extension;
+import sun.security.x509.SubjectAlternativeNameExtension;
+import sun.security.x509.X500Name;
+import sun.security.x509.X509CertImpl;
+import sun.security.x509.X509CertInfo;
+
 /**
  * Basic {@link SslEngineSource} for testing. The {@link SSLContext} uses
  * self-signed certificates that are generated lazily if the given key store
  * file doesn't yet exist.
  */
 public class SelfSignedSslEngineSource implements SslEngineSource {
-    private static final Logger LOG = LoggerFactory
-            .getLogger(SelfSignedSslEngineSource.class);
+    private static final Logger LOG = LoggerFactory.getLogger(SelfSignedSslEngineSource.class);
 
     private static final String ALIAS = "littleproxy";
     private static final String PASSWORD = "Be Your Own Lantern";
     private static final String PROTOCOL = "TLS";
+    private static final String KEY_STORE_TYPE = "jks";
+    private static final String SIG_ALG_NAME = "SHA256withRSA";
+    
+    
     private final File keyStoreFile;
     private final boolean trustAllServers;
     private final boolean sendCerts;
 
     private SSLContext sslContext;
 
+	private KeyStore keyStore;
+	
+	static ConcurrentHashMap<String, SoftReference<SSLContext>> sslContextCache = new ConcurrentHashMap<String, SoftReference<SSLContext>>();
```

Comment:
We even include guava already, so building an evicting map is very easy. Something like this:
```java
ConcurrentMap<String, SslContext> evictingMap = CacheBuilder.newBuilder()
  .expireAfterAccess(5, TimeUnit.MINUTES)
  .concurrencyLevel(16)
  .softValues()
  .build()
  .asMap();
```
The TTL and concurrency level will have to be tuned. I wonder if it would be desireable to be able to specify and/or change the concurrency level as a bootstrap parameter, since high-throughput scenarios may get bogged down locking the map.